karate-core/src/main/java/com/intuit/karate/http/HttpRequestBuilder.java

```diff
@@ -263,32 +258,29 @@ public HttpRequestBuilder path(String path) {
         return this;
     }
 
-    private String getPath() {
-        String temp = "";
-        if (paths == null) {
-            return temp;
-        }
-        for (String path : paths) {
-            if (path.startsWith("/")) {
-                path = path.substring(1);
+    private URI getUri() {
+        try {
+            URIBuilder builder = url == null ? new URIBuilder() : new URIBuilder(url);
+            if (params != null) {
+                params.forEach((key, values) -> values.forEach(value -> builder.addParameter(key, value)));
             }
-            if (!temp.isEmpty() && !temp.endsWith("/")) {
-                temp = temp + "/";
+            if (paths != null) {
+                paths.forEach(path -> {
+                    if (path.startsWith("/")) {
+                        logger.warn("Path segment: '{}' starts with a '/', this is probably a mistake. The '/' character will be escaped and sent to the remote server as '%2F'. " +
+                                "If you want to include multiple paths please separate them using commas. Ie. 'hello', 'world' instead of '/hello/world'.", path);
+                    }
+                });
+                // merge paths from the supplied url with additional paths supplied to this builder
+                List<String> merged = Stream.of(builder.getPathSegments(), paths)
+                        .flatMap(List::stream)
+                        .collect(Collectors.toList());
+                builder.setPathSegments(merged);
```

Comment:
Single level of abstraction:

You could probably organize this like so (forgive any syntax errors. Wrote this in the browser):

```
try {
    URIBuilder builder = createBuilder(url, params);
    if (Objects.nonNull(paths)) {
        builder.setPathSegments(createSegments(builder, paths));
    }

    // Rest of the code
}

private List<String> createSegments(URIBuilder builder, List<String> paths) {
    paths.stream()
        .filter(path -> path.startsWith("/"))
        .forEach(this::warnPathSegment);
    
    return Stream.of(builder.getPathSegments(), paths)
        .flatMap(List::stream)
        .collect(Collectors.toList());
}

private void warnPathSegment(String path) {
    logger.warn("Path segment: '{}' starts with a '/', this is probably a mistake. The '/' character will be escaped and sent to the remote server as '%2F'. " +
                                "If you want to include multiple paths please separate them using commas. Ie. 'hello', 'world' instead of '/hello/world'.", path);
}

// Not sure the type of params
private URIBuilder createBuilder(String url, List<K, V> params) {
    URIBuilder builder = Objects.isNull(url) ? new URIBuilder() : new URIBuilder(url);
    if (Objects.nonNull(params)) {
        params.forEach((key, values) -> values.forEach(value -> builder.addParameter(key, value)));
    }
    return builder;
}
```

Reply:
thanks nice, I like that, thanks @lodenrogue